lockin/src/lib.rs

```diff
@@ -0,0 +1,973 @@
+    // one that has not yet been set).
+    pub sequences_old: i32,
+}
+
+impl TimeStamp {
+    // Initialize a new TimeStamp.
+    pub fn new() -> Self {
+        TimeStamp {
+            count: 0,
+            sequences_old: -1,
+        }
+    }
+
+    /// Increments `sequences_old` if TimeStamp is valid. This is
+    /// called at the end of processing a sequence of ADC samples.
+    pub fn new_sequence(&mut self) {
+        if self.sequences_old != -1 {
+            self.sequences_old += 1;
+        }
+    }
+
+    /// Returns true if TimeStamp is valid.
+    pub fn is_valid(&self) -> bool {
+        self.sequences_old != -1
+    }
+
+    /// Set a new count value for TimeStamp. This also changes
+    /// `sequences_old` to 0 to indicate the TimeStamp belongs to the
+    /// current processing sequence.
+    ///
+    /// # Arguments
+    ///
+    /// * `newval` - New count value.
+    pub fn new_count(&mut self, newval: u16) {
+        self.count = newval as u32;
+        self.sequences_old = 0;
+    }
+}
+
+/// Compute the reference signal phase increment for each ADC sample
+/// and use them to demodulate the ADC inputs.
+macro_rules! prefilt_no_decimate {
+    ( $adc_samples:expr, $ref_tstamps:expr, $valid_tstamps:expr, $phi:expr, $tadc:expr, $fscale:expr, $tstamps_mem:expr) => {{
+        record_new_tstamps($ref_tstamps, $valid_tstamps, $tstamps_mem);
+        let ts_valid_count = tstamps_valid_count($tstamps_mem);
+
+        if ts_valid_count < 2 {
+            return ([0.; OUTPUT_BUFFER_SIZE], [0.; OUTPUT_BUFFER_SIZE]);
+        }
+
+        let thetas = adc_phases(
+            $ref_tstamps[0] as u32,
+            $tstamps_mem,
+            $phi,
+            $fscale,
+            $tadc,
+        );
+        let mut sines: [f32; SAMPLE_BUFFER_SIZE] = [0.; SAMPLE_BUFFER_SIZE];
+        let mut cosines: [f32; SAMPLE_BUFFER_SIZE] = [0.; SAMPLE_BUFFER_SIZE];
+        for i in 0..SAMPLE_BUFFER_SIZE {
+            sines[i] = sin(thetas[i]);
+            cosines[i] = cos(thetas[i]);
+        }
+
+        increment_tstamp_sequence($tstamps_mem);
+        demod($adc_samples, sines, cosines)
+    }};
+}
+
+/// Unfiltered in-phase and quadrature signals.
+///
+/// * `adc_samples` - ADC samples.
+/// * `ref_tstamps` - Slow external reference clock counter values.
+/// * `valid_tstamps` - Number of valid timestamps in `ref_tstamps`.
+/// * `phi` - Demodulation phase offset specified as a number of counts
+/// of the internal clock period.
+/// * `tadc` - Number of internal clock periods in one ADC sampling
+/// period.
+/// * `fscale` - Scaling factor for the demodulation frequency. For
+/// instance, 2 would demodulate with the first harmonic of the reference
+/// frequency.
+/// * `tstamps_mem` - Last two external reference timestamps (i.e., recorded
+/// values of `ref_tstamps`.)
+///
+/// # Latency (ADC batch size = 16): 5.6us
+///
+/// * 2.0us to compute ADC phase values
+/// * 3.4us to compute sin and cos
+pub fn prefilt(
```

Comment:
I think this would actually benefit a lot by creating a `DspFilter` object and then `processing()` ADC samples to generate I/Q.

Sample API:
```rust
pub struct DspFilter {
    phase_offset: u32,
    adc_cycles: u32,
    frequency_scale: u32,
    timestamps: [TimeStamp; 2].
}

impl DspFilter {
    pub fn prefilter(adc_samples: &[i16], timestamps: &[u16]) -> Result<([f32; OUTPUT_BUFFER_SIZE], [f32; OUTPUT_BUFFER_SIZE]), Error> {
        // ...
    }

    // Etcetera
}
```

I believe Rust may be able to analyze things better and recognize that certain parameters are actually `const` as well.